sdk/messaging/azeventhubs/processor_load_balancer.go

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package azeventhubs
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/rand"
+	"time"
+)
+
+type processorLoadBalancer struct {
+	checkpointStore             CheckpointStore
+	details                     consumerClientDetails
+	strategy                    ProcessorStrategy
+	partitionExpirationDuration time.Duration
+}
+
+func newProcessorLoadBalancer(checkpointStore CheckpointStore, details consumerClientDetails, strategy ProcessorStrategy, partitionExpiration time.Duration) *processorLoadBalancer {
+	return &processorLoadBalancer{
+		checkpointStore:             checkpointStore,
+		details:                     details,
+		strategy:                    strategy,
+		partitionExpirationDuration: partitionExpiration,
+	}
+}
+
+type loadBalancerInfo struct {
+	// current are the partitions that _we_ own
+	current []Ownership
+
+	// unownedOrExpired partitions either had no claim _ever_ or were once
+	// owned but the ownership claim has expired.
+	unownedOrExpired []Ownership
+
+	// aboveMax are ownerships where the specific owner has too many partitions
+	// it contains _all_ the partitions for that particular consumer.
+	aboveMax []Ownership
+
+	// maxAllowed is the maximum number of partitions a consumer should have
+	// If partitions do not divide evenly this will be the "theoretical" max
+	// with the assumption that this particular consumer will get an extra
+	// partition.
+	maxAllowed int
+
+	// extraPartitionPossible is true if the partitions cannot split up evenly
+	// amongst all the known consumers.
+	extraPartitionPossible bool
+
+	raw []Ownership
+}
+
+// loadBalance calls through to the user's configured load balancing algorithm.
+func (lb *processorLoadBalancer) LoadBalance(ctx context.Context, partitionIDs []string) ([]Ownership, error) {
```

Comment:
Is there some prior art here, i.e. are the algorithms adapted from some other implementation or is this all new stuff?

Reply:
Prior art - these are the two algorithms we have in the other track 2 libraries.